src/bin/pgcopydb/pgsql_timeline.c

```diff
@@ -0,0 +1,363 @@
+				current->tli,
+				(uint32_t) (current->begin >> 32),
+				(uint32_t) current->begin);
+	}
+
+	if (connIsOurs)
+	{
+		(void) pgsql_finish(pgsql);
+	}
+
+	return true;
+}
+
+
+/*
+ * parseIdentifySystemResult parses the result from a replication query
+ * IDENTIFY_SYSTEM, and fills the given IdentifySystem structure.
+ */
+static void
+parseIdentifySystemResult(void *ctx, PGresult *result)
+{
+	IdentifySystemResult *context = (IdentifySystemResult *) ctx;
+
+	if (PQnfields(result) != 4)
+	{
+		log_error("Query returned %d columns, expected 4", PQnfields(result));
+		context->parsedOk = false;
+		return;
+	}
+
+	if (PQntuples(result) == 0)
+	{
+		log_sql("parseIdentifySystem: query returned no rows");
+		context->parsedOk = false;
+		return;
+	}
+	if (PQntuples(result) != 1)
+	{
+		log_error("Query returned %d rows, expected 1", PQntuples(result));
+		context->parsedOk = false;
+		return;
+	}
+
+	/* systemid (text) */
+	char *value = PQgetvalue(result, 0, 0);
+	if (!stringToUInt64(value, &(context->system->identifier)))
+	{
+		log_error("Failed to parse system_identifier \"%s\"", value);
+		context->parsedOk = false;
+		return;
+	}
+
+	/* timeline (int4) */
+	value = PQgetvalue(result, 0, 1);
+	if (!stringToUInt32(value, &(context->system->timeline)))
+	{
+		log_error("Failed to parse timeline \"%s\"", value);
+		context->parsedOk = false;
+		return;
+	}
+
+	/* xlogpos (text) */
+	value = PQgetvalue(result, 0, 2);
+	strlcpy(context->system->xlogpos, value, PG_LSN_MAXLENGTH);
+
+	/* dbname (text) Database connected to or null */
+	if (!PQgetisnull(result, 0, 3))
+	{
+		value = PQgetvalue(result, 0, 3);
+		strlcpy(context->system->dbname, value, NAMEDATALEN);
+	}
+
+	context->parsedOk = true;
+}
+
+
+/*
+ * parseTimelineHistoryResult parses the result of the TIMELINE_HISTORY
+ * replication command.
+ */
+static void
+parseTimelineHistoryResult(void *ctx, PGresult *result)
+{
+	TimelineHistoryResult *context = (TimelineHistoryResult *) ctx;
+
+	if (PQnfields(result) != 2)
+	{
+		log_error("Query returned %d columns, expected 2", PQnfields(result));
+		context->parsedOk = false;
+		return;
+	}
+
+	if (PQntuples(result) == 0)
+	{
+		log_sql("parseTimelineHistoryResult: query returned no rows");
+		context->parsedOk = false;
+		return;
+	}
+
+	if (PQntuples(result) != 1)
+	{
+		log_error("Query returned %d rows, expected 1", PQntuples(result));
+		context->parsedOk = false;
+		return;
+	}
+
+	/* filename (text) */
+	char *value = PQgetvalue(result, 0, 0);
+	strlcpy(context->filename, value, sizeof(context->filename));
+
+	/* content (bytea) */
+	value = PQgetvalue(result, 0, 1);
+
+	if (strlen(value) >= sizeof(context->content))
+	{
+		log_error("Received a timeline history file of %zu bytes, "
+				  "pgcopydb is limited to files of up to %zu bytes.",
+				  strlen(value),
+				  sizeof(context->content));
+		context->parsedOk = false;
+	}
+	strlcpy(context->content, value, sizeof(context->content));
+
+	context->parsedOk = true;
```

Comment:
Can we fetch the value length and dynamically allocate a buffer of the proper size here, using `calloc`?

Reply:
Apparently we can use (PQgetlength)[https://www.postgresql.org/docs/current/libpq-exec.html#LIBPQ-PQGETLENGTH] for that.

Reply:
Or just `strdup` ;-)